File: samples/Orleans.Dataflow.Samples.FSharp/Junctions.fs
namespace Orleans.Dataflow.Samples.FSharp

open System.Threading
open System.Threading.Tasks
open Orleans.Dataflow.FSharp
open Orleans.Dataflow.Samples

// Orleans.Dataflow itself is deliberately not opened: see the note in FirstPipeline.fs.

/// <summary>One stream broadcast into two branches, each with a result of its own.</summary>
/// <remarks>
/// <para>
/// A branch is a flow that ends in a terminal, and a junction call is what turns a list of them into a
/// closed graph. Every element reaches every branch, so the two counts below are two readings of the same
/// twelve orders rather than a partition of them — which is the difference between a broadcast and the
/// balance and partition junctions beside it.
/// </para>
/// <para>
/// Each branch names the slot its result resolves under, so one run answers two questions. The slots are in
/// the author's hand before the junction call is written, which is why the junction call itself answers only
/// the graph.
/// </para>
/// </remarks>
[<RequireQualifiedAccess>]
module Junctions =

    /// <summary>What an order has to be worth to count as large.</summary>
    let private large = 50m

    /// <summary>Authors the broadcast, runs it, and reports both branches' results.</summary>
    /// <param name="sample">The run this scenario belongs to.</param>
    /// <param name="cancellationToken">The whole run's budget.</param>
    /// <returns>The graph's fingerprint and the two counts.</returns>
    [<CompiledName("RunAsync")>]
    let runAsync (sample: SampleRun) (cancellationToken: CancellationToken) : Task<ScenarioOutcome> =
        task {
            let orders = SampleOrders.Take(sample.Scale.Pick(full = 12, smokeSize = 6))

            let largeBranch, largeSlot =
                Flow.filter (fun (document: OrderDocument) -> document.Amount >= large)
                |> Branch.toResult "large" Sink.count

            let northBranch, northSlot =
                Flow.filter (fun (document: OrderDocument) -> document.Region = "north")
                |> Branch.toResult "north" Sink.count

            let graph =
                Source.ofSeq orders
                |> Source.map OrderDocument.ofEvent
                |> Source.broadcastTo [ largeBranch; northBranch ]

            let host = Orleans.Dataflow.LocalDataflowHost()
            use! run = host.MaterializeAsync(graph, cancellationToken)
            let! largeOrders = run |> Run.value largeSlot cancellationToken
            let! northOrders = run |> Run.value northSlot cancellationToken

            do! run.Completion

            return
                ScenarioOutcome.Of(
                    [ GraphReading.Of("main", graph) ],
                    [ Observation.Of("orders-broadcast", orders.Count)
                      Observation.Of("orders-worth-50-or-more", largeOrders)
                      Observation.Of("orders-from-the-north", northOrders) ]
                )
        }

File: samples/Orleans.Dataflow.Samples/CSharp/Junctions.cs
namespace Orleans.Dataflow.Samples.CSharp;

/// <summary>
/// One stream broadcast into two branches, each with a result of its own.
/// </summary>
/// <remarks>
/// <para>
/// A branch is a flow that ends in a terminal, and a junction call is what turns a list of them into a
/// closed graph. Every element reaches every branch, so the two counts below are two readings of the same
/// orders rather than a partition of them — which is the difference between a broadcast and the balance and
/// partition junctions beside it.
/// </para>
/// <para>
/// A broadcast asks every leg for room before it pulls, so a branch that stops consuming holds all of them
/// up. That is the bounded memory this junction buys: nothing anywhere accumulates on behalf of a slow leg.
/// </para>
/// </remarks>
internal static class Junctions
{
    /// <summary>What an order has to be worth to count as large.</summary>
    private const decimal Large = 50m;

    /// <summary>Authors the broadcast, runs it, and reports both branches' results.</summary>
    /// <param name="sample">The run this scenario belongs to.</param>
    /// <param name="cancellationToken">The whole run's budget.</param>
    /// <returns>The graph's fingerprint and the two counts.</returns>
    internal static async Task<ScenarioOutcome> RunAsync(SampleRun sample, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(sample);

        IReadOnlyList<OrderEvent> orders = SampleOrders.Take(sample.Scale.Pick(full: 12, smokeSize: 6));

        Branch<OrderDocument> largeBranch = Flow.For<OrderDocument>()
            .Where(document => document.Amount >= Large)
            .To(Sink.Count<OrderDocument>(), "large", out ResultSlot<long> largeSlot);

        Branch<OrderDocument> northBranch = Flow.For<OrderDocument>()
            .Where(document => document.Region == "north")
            .To(Sink.Count<OrderDocument>(), "north", out ResultSlot<long> northSlot);

        RunnableGraph graph = Source.From(orders)
            .Select(OrderDocument.FromEvent)
            .BroadcastTo(largeBranch, northBranch);

        await using RunHandle run = await new LocalDataflowHost().MaterializeAsync(graph, cancellationToken);

        long largeOrders = await run.GetValueAsync(largeSlot, cancellationToken);
        long northOrders = await run.GetValueAsync(northSlot, cancellationToken);

        await run.Completion;

        return ScenarioOutcome.Of(
            [GraphReading.Of("main", graph)],
            [
                Observation.Of("orders-broadcast", orders.Count),
                Observation.Of("orders-worth-50-or-more", largeOrders),
                Observation.Of("orders-from-the-north", northOrders),
            ]);
    }
}
